File: GA/ga_toolbox/selection_uniq.m
function [YY1, YY2] = selection_uniq(P1,B,p,r,op,LR,UR)
% P1: population
% B : fitness value
% p : population size
% s : select top s chromosomes

% YY1 : selected population
% YY2 : the trun fitness of YY1

%f = S(find(rand<cumsum(w),1,'first'))

% trun min. to max.
B = op*B;
dim = size(P1,2);

%first kill inf
index = isinf(B);
B(index) = [];
P1(index,:) = [];
index = isnan(B);
B(index) = [];
P1(index,:) = [];

% check firt the unique, keep unique
[P1,ip,~] = unique(P1,'rows');
B = B(ip);


% sort from high to low , get the best
[E,index] = sort(B,'descend'); % E all sorted cost
P1 = P1(index,:); % all sorted pop.

% recheck the length, min shold be p
len_counter = length(E); % total length
if len_counter < p
    new_len = p-len_counter;
    P1 = [P1;population(new_len,dim,LR,UR)];
    E = [E,nan(1,new_len)];
end
% select best
% randon draw the rest
if len_counter > p
    P1_best = P1(1:r,:);
    E_best = E(1:r);
    P2 = P1(r+1:end,:);
    E2 = E(r+1:end);
    index = randperm(length(E2));
    P2 =P2(index,:);
    E2 =E2(index);
    P2 = P2(1:p-r,:);
    E2 = E2(1:p-r);
    P1 = [P1_best;P2];
    E = [E_best,E2];
    YY1 = P1(1:p,:);
    YY2 = op*E(1:p);  % turn back
else
    YY1 = P1(1:p,:);
    YY2 = op*E(1:p);  % turn back
end


end


  
    
 

File: GA/ga_toolbox/selection_uniq.py
import numpy as np
from GA.ga_toolbox.population import population


def selection_uniq(P1, B, p, r, op, LR, UR):
    """
    Select p unique solutions, keeping the top-r best and randomly sampling
    the rest from the remaining pool.

    Parameters
    ----------
    P1 : np.ndarray  [n_pop x n_parameter]  population
    B  : np.ndarray  [n_pop,]               fitness values
    p  : int         desired output population size
    r  : int         number of top solutions guaranteed to be kept
    op : int         -1 → select minimum, +1 → select maximum
    LR : array-like  lower boundary
    UR : array-like  upper boundary

    Returns
    -------
    YY1 : np.ndarray  [p x n_parameter]  selected population
    YY2 : np.ndarray  [p,]               fitness of YY1
    """
    # Turn minimisation into maximisation if necessary
    B = op * B.copy()
    dim = P1.shape[1]

    # Remove inf entries
    index = np.isinf(B)
    B  = B[~index]
    P1 = P1[~index, :]

    # Remove nan entries
    index = np.isnan(B)
    B  = B[~index]
    P1 = P1[~index, :]

    # Keep only unique rows
    P1, ip = np.unique(P1, axis=0, return_index=True)
    B = B[ip]

    # Sort from high to low — best first
    index = np.argsort(B)[::-1]
    E  = B[index]
    P1 = P1[index, :]

    # Recheck length; pad with random solutions if fewer than p remain
    len_counter = len(E)
    if len_counter < p:
        new_len = p - len_counter
        P1 = np.vstack([P1, population(new_len, dim, LR, UR)])
        E  = np.concatenate([E, np.full(new_len, np.nan)])

    # Select best r, then randomly sample the rest
    if len_counter > p:
        P1_best = P1[:r, :]
        E_best  = E[:r]

        P2 = P1[r:, :]
        E2 = E[r:]

        index = np.random.permutation(len(E2))
        P2 = P2[index, :]
        E2 = E2[index]

        P2 = P2[:p - r, :]
        E2 = E2[:p - r]

        P1 = np.vstack([P1_best, P2])
        E  = np.concatenate([E_best, E2])

    YY1 = P1[:p, :]
    YY2 = op * E[:p]   # turn back to original sign

    return YY1, YY2